lib/web_catalog/api_extractor.es6.js

```diff
@@ -140,511 +242,428 @@ foam.CLASS({
-                // Walk the prototype chain of <current object>.prototype until:
-                // (1) the prototype's class name changes (i.e., prototype
-                //     belongs to another API; OR
-                // (2) Object.prototype is reached (this is necessary because
-                //     the current class name may be "Object").
-                const objectPrototypePropertyId = og.lookup('Object.prototype');
-                let prototypePropertyClassName =
-                    this.getClassName_(prototypePropertyId, og);
-                while (!og.isType(prototypePropertyId) &&
-                       prototypePropertyId !== objectPrototypePropertyId &&
-                       prototypePropertyClassName === className) {
-                  prototypePropertyProps = prototypePropertyProps.concat(
-                      this.getObjectProperties_(og, prototypePropertyId));
-                  prototypePropertyId = og.getPrototype(prototypePropertyId);
-                  prototypePropertyClassName =
-                      this.getClassName_(prototypePropertyId, og);
+        }
+
+        // Gather APIs from libraries function/prototype pairs.
+        for (const id of allOgIds) {
+          const isCtor = this.fns_.hasOwnProperty(id);
+          const isLib = this.libs_.hasOwnProperty(id);
+          const isCtorProto = this.protos_.hasOwnProperty(id);
+
+          foam.assert(!((isCtor && isLib) || (isCtor && isCtorProto) ||
+                        (isLib && isCtorProto)),
+                      'ApiExtractor: Object with multiple personalities');
+
+          // Functions are interfaces, not prototypes. Just store properties and
+          // carry on.
+          if (isCtor) {
+            this.storeAPIsFromCtorOrLib_(og, id);
+          } else if (isLib) {
+            this.storeAPIsFromCtorOrLib_(og, id);
+            // Treat library as interface where Ctor = Ctor.prototype.
+            const ctorId = id;
+            const ctorNames = this.libs_[id];
+            let protoId = id;
+            let protoNames = this.getClassNames_(og, protoId);
+            this.storeProtoAPIsForCtor_(og, ctorId, ctorNames, protoId,
+                                        protoNames);
+          } else if (isCtorProto) {
+            // This is a "Foo.prototype", or a library object. Store its
+            // properties under "Foo", and pull up any of its prototype's
+            // properties, so long as those prototypes are not some
+            // "Bar.prototype".
+            const ctorIds = this.protos_[id];
+            for (const ctorId of ctorIds) {
+              const ctorNames = this.getFunctionNames_(og, ctorId);
+              let protoId = id;
+              let protoNames = this.getClassNames_(og, protoId);
+              this.storeProtoAPIsForCtor_(og, ctorId, ctorNames, protoId,
+                                          protoNames);
+            }
+          } // else: This is an instance or a primitive value.
```

Comment:
Sort of along the same line, adding an else branch that populates a list of non-constructor/library/prototype would mean less work later, because the filtering is done.